### PR TITLE
New DataBaseManagment

### PR DESCRIPTION
**Pull request by dominikpalatynski**
DataBaseManager with two methods fetching data and adding data to DataBase + Unit tests

### CODE REVIEW

File: source/Backend/DataBase/DataBaseConnection.hpp [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  DataBaseConnection(const char *dbName);
+
+  ~DataBaseConnection();
+
+  void executeQuery(const std::string &query) override;
+  std::vector<std::string> executeQueryAndGetData(
+      const std::string &query) override;
+
+ private:
+  void checkResult(int result, char *errMsg);
```
Comment (emttiew) [inline source/Backend/DataBase/DataBaseConnection.hpp]: Move this function to unnamed namespace in cpp file. It doesnt need to be a part of the class interface.

File: source/Backend/DataBase/QueryOperationBuilder.cpp [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+}
+
+std::string buildSelectQuery(const common::DataBaseRequest &requestData,
+                             const std::string &tableName) {
+  return std::format("SELECT * FROM {} WHERE {}='{}';", tableName,
+                     requestData.primaryKey_.first,
+                     requestData.primaryKey_.second);
+}
+std::unordered_map<common::DataBaseRequest::TableType, std::string> tableNames =
+    {{common::DataBaseRequest::TableType::Users, common::users},
```
Comment (emttiew) [inline source/Backend/DataBase/QueryOperationBuilder.cpp]:
```cpp
using TableType = common::DataBaseRequest::TableType;
```

File: source/Backend/DataBase/QueryOperationBuilder.cpp [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+};
+auto updateCallback = [](const auto &requestData, const auto &tableName) {
+  return buildUpdateQuery(requestData, tableName);
+};
+auto deleteCallback = [](const auto &requestData, const auto &tableName) {
+  return buildDeleteQuery(requestData, tableName);
+};
+auto selectCallback = [](const auto &requestData, const auto &tableName) {
+  return buildSelectQuery(requestData, tableName);
+};
```
Comment (emttiew) [inline source/Backend/DataBase/QueryOperationBuilder.cpp]: I think you don't need these lambdas. You can just simply call build function
```cpp
{QueryOperationBuilder::OperationType::Insert, buildInsertQuery}
```

File: source/Backend/DataBase/QueryOperationBuilder.hpp [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+ private:
+  std::string buildInsertQuery(const common::DataBaseRequest &requestData,
+                               const std::string &tableName);
+  std::string buildUpdateQuery(const common::DataBaseRequest &requestData,
+                               const std::string &tableName);
+  std::string buildDeleteQuery(const common::DataBaseRequest &requestData,
+                               const std::string &tableName);
+  std::string buildSelectQuery(const common::DataBaseRequest &requestData,
+                               const std::string &tableName);
```
Comment (emttiew) [inline source/Backend/DataBase/QueryOperationBuilder.hpp]: Remove this whole section as it's no longer needed.

Comment (dominikpalatynski) [reply source/Backend/DataBase/QueryOperationBuilder.hpp]: thats right, I didn't knew if that was even possible :)

File: source/Backend/DataBase/QueryOperationBuilder.hpp [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+  class QueryOperationBuilder : public interface::IQueryOperationBuilder
+  {
+  public:
+    using QueryBuilders = std::unordered_map<
+        common::DataBaseRequest::OperationType,
+        std::function<std::string(const common::DataBaseRequest &,
+                                  const std::string &)>>;
+    using OperationType = common::DataBaseRequest::OperationType;
+    using TableType = common::DataBaseRequest::TableType;
```
Comment (emttiew) [inline source/Backend/DataBase/QueryOperationBuilder.hpp]: If this is only used in cpp file it is better to leave it there :)

Comment (dominikpalatynski) [reply source/Backend/DataBase/QueryOperationBuilder.hpp]: You right, I will put this in to anonymous namespace

**Review by emttiew — APPROVED** [review]

**Review by skwarq — APPROVED** [review]

**Comment by ziobron** [comment]
Your PR was merged!
🏅 2 XP granted. Thanks @dominikpalatynski for making course materials better!

**Pull request merged**